### PR TITLE
APEXCORE-734 StramLocalCluster may not terminate properly

### PR DESCRIPTION
**Pull request by vrozov**
@PramodSSImmaneni @tweise Please review

### CODE REVIEW

**Comment by vrozov** [comment]
@PramodSSImmaneni @tweise Please review

**Comment by vrozov** [comment]
@PramodSSImmaneni @tweise Please review

**Comment by pramodin** [comment]
Will look into it

**Comment by pramodin** [comment]
Can you rebase

File: engine/src/main/java/com/datatorrent/stram/StramLocalCluster.java [hunk 1]
```diff
@@ -463,103 +463,123 @@ public void run()
...
-    LOG.info("Application finished.");
-    if (!perContainerBufferServer) {
-      bufferServer.stop();
-      StreamingContainer.eventloop.stop();
+      if (eventLoopThread != null) {
+        try {
+          StreamingContainer.eventloop.stop();
+          eventLoopThread.join(1000);
+        } catch (InterruptedException ie) {
+          LOG.debug("Sleep interrupted", ie);
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/StramLocalCluster.java]: Should this be debug or warn?

**Comment by vrozov** [comment]
Now it is not necessary to rebase unless there is a conflict. Please use "Rebase and merge" option on github.

**Comment by pramodin** [comment]
Please take a look at the comment as well and address it when you get a chance. 

Also, IMO, not having the conflict should not be the only criteria as there could be other logical problems after rebase and merge. I prefer the general guideline to committers not be an automatic rebase and merge but rather to make an assessment based on the situation like the changes made in the meantime and if needed do a rebase locally, check and make the appropriate decision. If needed go back to the contributor. 

That being said, this particular PR likely can be rebased and merged from github as other recent changes or PRs in consideration that could get merged before this PR don't affect this.

**Comment by vrozov** [comment]
The debug level is intentional. The StramLocalCluster main thread may be interrupted (please see description in JIRA). Warning is logged in case an event loop thread or container threads do not terminate, the extra warning is not necessary.

I guess we need to discuss the policy for the rebase on dev@apex. For smaller changes a committer does not have to require rebase and should be in a position to merge after review is done, IMO. If changes are limited to a file or a method that were not recently changed, the rebase should not be required as well. Squashing commits by a contributor is a must.

**Comment by tweise** [comment]
If CI fails as result of automatic rebase then the committer performing the push needs to be responsible. It is possible that a merge succeeds but the build doesn't pass.

@vrozov please take the discussion to the mailing list, as well as the one from the other PR.

**Comment by vrozov** [comment]
Jenkins CI runs against the merge, not against the original commit.

**Pull request merged**